sstables/storage.cc

```diff
@@ -800,6 +800,9 @@ std::unique_ptr<sstables::storage> make_storage(sstables_manager& manager, const
             if (s_opts.is_s3_type()) {
                 return std::make_unique<sstables::s3_storage>(manager.get_endpoint_client(os.endpoint), os.bucket, os.location, os.abort_source);
```

Comment:
Hm... AFAICS this doesn't get converted into `make_unique<object_storage_base>("S3", ...)` which makes me think that wrapping s3_storage to inherit from object_storage_base is incomplete, is it?

Reply:
> The issue is make_data_or_index_source, which for the GS (or base) variant ... is implemented as always just wrapping a data_source objec

That's exactly my question. Why is `object_storage_base::make_data_or_index_source()` implemented like that? 
Why can't it just call `object_storage_client::make_data_or_index_source()` and let GS version do maybe_wrap_source on its download source and S3 version take its detour (or fallback to base implementation as it does now if offset == 0)?

Maybe you want to push S3 client further to a state where it's no longer "challenged" and can thus enjoy this PR's base make_data_or_index_source implementation? Or is there something in GS that's inherently missing in S3 that doesn't allow it being such? Or something else?

Reply:
The S3 specialisation exists only in wait for S3 to have proper seek. Once it does, there is no reason for it. I did not want to modify S3 proper (except for some aux methods being moved) in this patch set. 

The object_storage_client cannot do wrapping, as this is on storage level. The idea is that the object_storage_client does not know about sstables as such, and is agnostic to the further chaining of sources/sinks/files.